src/py21cmfast/src/GenerateICs.c

```diff
@@ -47,44 +47,82 @@ int ComputeInitialConditions(unsigned long long random_seed, struct UserParams *
      Author: Andrei Mesinger
      Date: 9/29/06
      */
-
+    
     // Makes the parameter structs visible to a variety of functions/macros
     // Do each time to avoid Python garbage collection issues
     Broadcast_struct_global_PS(user_params,cosmo_params);
     Broadcast_struct_global_UF(user_params,cosmo_params);
-
+    
     fftwf_plan plan;
 
     char wisdom_filename[500];
 
     unsigned long long ct;
-    int n_x, n_y, n_z, i, j, k, ii;
+    int n_x, n_y, n_z, i, j, k, ii, thread_num;
     float k_x, k_y, k_z, k_mag, p, a, b, k_sq;
     double pixel_deltax;
     float p_vcb, vcb_x, vcb_y, vcb_z;
 
     float f_pixel_factor;
 
-    gsl_rng * r;
+    gsl_rng * r[user_params->N_THREADS];
+    
+    omp_set_num_threads(user_params->N_THREADS);
+    fftwf_init_threads();
+    fftwf_plan_with_nthreads(user_params->N_THREADS);
 
     // ************  INITIALIZATION ********************** //
 
     // Removed all references to threads as 21CMMC is always a single core implementation
 
+    int checker;
+    
+    checker = 0;
     // seed the random number generators
-    r = gsl_rng_alloc(gsl_rng_mt19937);
-    gsl_rng_set(r, random_seed);
+    for (thread_num = 0; thread_num < user_params->N_THREADS; thread_num++){
+        switch (checker){
```

Comment:
Rather than cycle through the random number generators in this linear way, which could lead to some residual correlation between threads that use the same generator, a safer way to do this is to generate the individual thread seeds with a random number generator from the initial seed and shuffle them. See https://rpubs.com/Jouni_Helske/225931 for details. 

My rough implementation in the old 21cmFAST looked like this:
```
int num_int = INT_MAX/16;
unsigned int *many_ints = (unsigned int *)malloc((size_t)(num_int*sizeof(unsigned int))); // some large number of possible integers
for (i=0; i<num_int; i++) {
    many_ints[i] = i;
}
gsl_rng * rseed = gsl_rng_alloc(gsl_rng_mt19937);
gsl_rng_set(rseed, RANDOM_SEED);
unsigned int seeds[NUMCORES];
gsl_ran_choose(rseed, seeds, NUMCORES, many_ints, num_int, sizeof(unsigned int));
gsl_ran_shuffle(rseed, seeds, NUMCORES, sizeof(unsigned int));
for (thread_num=0; thread_num < NUM_RNG_THREADS; thread_num++) {
    r[thread_num] = gsl_rng_alloc(gsl_rng_ranlxs2);
    gsl_rng_set(r[thread_num],seeds[thread_num]);
    fprintf(stderr, "Thread #%i will use a ranlxs2 RNG seeded with %u\n", thread_num, seeds[thread_num]);
}

Reply:
Thanks @freddavies! I agree this is a nice way of dealing with this. I originally toyed with having each RNG being seeded by a random product of random numbers. 

I'm not sure what you mean by residual correlations between threads that use the same generator. Each thread uses its own differently seeded RNG so no generator has the same seed.

Reply:
The blog post shows that if the set of initial seeds is too simply structured (e.g., a linear function of thread ID, as done here), then there are residual correlations between the values generated by the various streams.  The first set of figures shows that there is a non-negligible correlation between the first random values drawn using the set of seeds from 1 to 100000 for a few common generators.

It's likely that this doesn't matter too much, but it's probably better to be safe than sorry. I can put in a PR later as @andreimesinger suggested.

Reply:
@freddavies ah, sorry. I didn't actually get a chance to look at the linked article.